### PR TITLE
[4379] New copy, cut & duplicate operations

### PR DESCRIPTION
**Pull request by aachavarria**
https://github.com/craftercms/craftercms/issues/4379

Adding policy to createFolderDialog;

### CODE REVIEW

File: ui/app/src/models/Site.ts [hunk 1]
```diff
@@ -113,3 +113,21 @@ export interface LegacySite {
   siteId: string;
   desc: string;
 }
+
+export interface Action {
+  type: string;
```
Comment (rart) [inline ui/app/src/models/Site.ts]:
```suggestion
  type: 'create' | 'blah' | 'yada';
```

File: ui/app/src/services/sites.ts [hunk 2]
```diff
@@ -66,3 +66,15 @@ export function updateSite(site: Site) {
 export function checkHandleAvailability(name: string) {
   return get(`/studio/api/1/services/api/1/site/exists.json?site=${name}`);
 }
+
+export function validateActionPolicy(site: string, action: Action): Observable<ContentValidationResult> {
+  return postJSON(`/studio/api/2/sites/${site}/policy/validate`, {
+    actions: [action]
+  }).pipe(pluck('response', 'results', '0'));
```
Comment (rart) [inline ui/app/src/services/sites.ts]: call the bulk one internally (don't duplicate)

File: ui/app/src/components/Dialogs/CreateFolderDialog.tsx [hunk 3]
```diff
@@ -113,29 +115,48 @@ function CreateFolderUI(props: CreateFolderUIProps) {
...
+      const parentPath = rename ? getParentPath(path) : path;
+      validateActionPolicy(site, {
+        type: rename ? 'RENAME' : 'CREATE',
+        target: `${parentPath}/${name}`
+      }).subscribe(({ allowed, modifiedValue }) => {
+        if (allowed) {
+          // TODO: add confirmation if modifiedValue?
+          let _name = modifiedValue ? modifiedValue.replace(`${parentPath}/`, '') : name;
+          if (rename) {
+            renameFolder(site, path, encodeURIComponent(_name)).subscribe(
```
Comment (rart) [inline ui/app/src/components/Dialogs/CreateFolderDialog.tsx]: encode on the service

File: ui/app/src/components/Dialogs/CreateFolderDialog.tsx [hunk 4]
```diff
@@ -113,29 +115,48 @@ function CreateFolderUI(props: CreateFolderUIProps) {
...
+                onRenamed?.({ path, name, rename });
+                dispatch(emitSystemEvent(folderRenamed({ target: path, oldName: value, newName: _name })));
+              },
+              (response) => {
+                setState({ inProgress: false, submitted: true });
+                dispatch(showErrorDialog({ error: response }));
+              }
+            );
+          } else {
+            createFolder(site, path, encodeURIComponent(_name)).subscribe(
```
Comment (rart) [inline ui/app/src/components/Dialogs/CreateFolderDialog.tsx]: encode on the service

File: ui/app/src/services/sites.ts [hunk 5]
```diff
@@ -84,3 +84,13 @@ export function update(site: Site): Observable<Site> {
...
+
+export function validateActionPolicy(site: string, action: Action): Observable<ContentValidationResult> {
+  return validateActionsPolicy(site, [action]).pipe(pluck('0'));
+}
+
+export function validateActionsPolicy(site: string, actions: Action[]): Observable<ContentValidationResult[]> {
+  return postJSON(`/studio/api/2/sites/${site}/policy/validate`, {
+    actions
+  }).pipe(pluck('response', 'results'));
+}
```
Comment (rart) [inline ui/app/src/services/sites.ts]:
```suggestion
export function validateActionPolicy(site: string, action: Action): Observable<ContentValidationResult>;
export function validateActionPolicy(site: string, actions: Action[]): Observable<ContentValidationResult[]>;
export function validateActionPolicy(site: string, action: Action | Action[]): Observable<ContentValidationResult | ContentValidationResult[]> {
  const multi = Array.isArray(action);
  const actions = multi ? action : [action];
  const toPluck = ['response', 'results', !multi && '0'].filter(Boolean);
  return postJSON(`/studio/api/2/sites/${site}/policy/validate`, {
    actions
  }).pipe(pluck(...toPluck));
}
```

File: ui/app/src/state/epics/content.ts [hunk 6]
```diff
@@ -178,13 +182,70 @@ const content = [
       })
     ),
   // endregion
-  // region Item Pasted
+  // region Duplicate with validation policy
+  (action$, state$: StateObservable<GlobalState>, { getIntl }) =>
+    action$.pipe(
+      ofType(duplicateWithValidationPolicy.type),
```
Comment (rart) [inline ui/app/src/state/epics/content.ts]: duplicateWithValidationPolicy => duplicateWithPolicyValidation

File: ui/app/src/state/epics/content.ts [hunk 7]
```diff
@@ -207,6 +268,42 @@ const content = [
           return NEVER;
         }
       })
+    ),
+  // endregion
+  // region Item Paste with validation policy
+  (action$, state$: StateObservable<GlobalState>, { getIntl }) =>
+    action$.pipe(
+      ofType(pasteItemWithValidationPolicy.type),
```
Comment (rart) [inline ui/app/src/state/epics/content.ts]: pasteItemWithValidationPolicy => pasteItemWithPolicyValidation

File: ui/app/src/utils/i18n-legacy.ts [hunk 8]
```diff
@@ -1478,3 +1478,15 @@ export const siteSuccessMessages = defineMessages({
...
+  itemPastePolicyConfirm: {
+    id: 'pastePolicy.confirm',
+    defaultMessage:
+      'The selected {action} target goes against site policies for the destination directory. • Original path: "{path}", • Suggested path is: "{modifiedPath}". Would you like to use the suggested path?'
+  },
+  itemPastePolicyError: {
+    id: 'pastePolicy.error',
+    defaultMessage: 'The selected copy|cut target goes against site policies for the destination directory.'
+  }
+});
```
Comment (rart) [inline ui/app/src/utils/i18n-legacy.ts]: Is this being used on legacy? Why is this here?

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/utils/i18n-legacy.ts (hunk 8 above)
Comment (aachavarria) [reply ui/app/src/utils/i18n-legacy.ts]: to avoid creating it on Epic, will move to the epic

File: ui/app/src/state/epics/content.ts [hunk 9]
```diff
@@ -41,21 +43,42 @@ import GlobalState from '../../models/GlobalState';
...
+
+export const sitePolicyMessages = defineMessages({
+  itemPastePolicyConfirm: {
+    id: 'pastePolicy.confirm',
+    defaultMessage:
+      'The selected {action} target goes against site policies for the destination directory. • Original path: "{path}", • Suggested path is: "{modifiedPath}". Would you like to use the suggested path?'
+  },
+  itemPastePolicyError: {
+    id: 'pastePolicy.error',
+    defaultMessage: 'The selected copy|cut target goes against site policies for the destination directory.'
```
Comment (rart) [inline ui/app/src/state/epics/content.ts]:
```suggestion
    defaultMessage: 'The selected {action} target goes against site policies for the destination directory.'
```

File: ui/app/src/utils/itemActions.ts [hunk 10]
```diff
@@ -313,12 +313,14 @@ export function generateSingleItemOptions(item: DetailedItem, permissions: Looku
         if (deleteItem) {
           _optionsA.push(menuOptions.delete);
         }
+        _optionsA.push(menuOptions.cut);
+        _optionsA.push(menuOptions.copy);
         if (type === 'taxonomy' || type === 'component') {
```
Comment (rart) [inline ui/app/src/utils/itemActions.ts]: taxonomy is not an existing system type

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/utils/itemActions.ts (hunk 10 above)
Comment (aachavarria) [reply ui/app/src/utils/itemActions.ts]: this was already changed on the PR https://github.com/craftercms/studio-ui/pull/2130

**Review by rart — APPROVED** [review]

**Review by rart — CHANGES_REQUESTED** [review]
[Build](https://travis-ci.com/github/craftercms/studio-ui/builds/214573324) is broken.

**Review by rart — APPROVED** [review]

**Comment by rart** [comment]
@aachavarria conflicts

**Review by rart — APPROVED** [review]
